Add Pulsar installation on OVHcloud doc

Sherlock113 opened the pull request:
@scraly @philippart-s PTAL. Thanks.

Sherlock113 committed:
Add Pulsar installation on OVHcloud doc

philippart-s reviewed (commented):
Hi, @Sherlock113 
Thanks for the writing of this great tutorial, I'll test it before accepting the PR.
Can you add in the prerequisites sections the minimal specs (RAM/CPU/number of nodes) needed to install both the operator and Pulsar please ?

Thanks

philippart-s commented:
Another idea: it will be great if you add a clean-up section at the end of the tutorial to explain how to remove pulsar, the operator and OLM?

philippart-s commented:
Another point: I think that adding a simple example of a command to test the pulsar installation will be useful too 😊

philippart-s commented:
@Sherlock113 after executing the following command: 
```bash
kubectl create -f https://raw.githubusercontent.com/streamnative/charts/master/examples/pulsar-operators/proxy.yaml
```

Only zookeepers pods are created:
```bash
kubectl get pods -n pulsar   
NAME              READY   STATUS    RESTARTS   AGE
zookeepers-zk-0   1/1     Running   0          78m
zookeepers-zk-1   1/1     Running   0          78m
zookeepers-zk-2   1/1     Running   0          78m
```

Do you have any idea of the reason?

Sherlock113 committed:
Update the doc

Sherlock113 committed:
Update the doc

Sherlock113 commented:
Hi @philippart-s Thank you for your review. I just updated the doc:

1. Added the minimum requirements. Could you please try to deploy operators and Pulsar again based on the minimum requirements? I think your case was caused by a lack of CPU and memory resources. I tried 3 D2-4 instances on OVHcloud and they worked properly to run Pulsar and operators.
2. Added the clean-up section.
3. Added a simple example to test the Pulsar cluster.

Let me know if you have other questions. Thanks.

Sherlock113 committed:
Update the doc

philippart-s commented:
Hi @Sherlock113,

Thanks for the updates.
Unfortunately I've recreated a D2-4 cluster with 3 nodes but only zookeepers pod were created.
You can see below my logs:
```bash
➜ kubectl get np
NAME                                            FLAVOR   AUTOSCALED   MONTHLYBILLED   ANTIAFFINITY   DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   MIN   MAX   AGE
nodepool-xxxxx-xxxxx-xxxx-xxxx-xxxx   d2-4     false        false           false          3         3         0            3           0     100   67m

➜ curl -sL https://github.com/operator-framework/operator-lifecycle-manager/releases/download/v0.24.0/install.sh | bash -s v0.24.0
customresourcedefinition.apiextensions.k8s.io/catalogsources.operators.coreos.com created
customresourcedefinition.apiextensions.k8s.io/clusterserviceversions.operators.coreos.com created
customresourcedefinition.apiextensions.k8s.io/installplans.operators.coreos.com created
customresourcedefinition.apiextensions.k8s.io/olmconfigs.operators.coreos.com created
customresourcedefinition.apiextensions.k8s.io/operatorconditions.operators.coreos.com created
customresourcedefinition.apiextensions.k8s.io/operatorgroups.operators.coreos.com created
customresourcedefinition.apiextensions.k8s.io/operators.operators.coreos.com created
customresourcedefinition.apiextensions.k8s.io/subscriptions.operators.coreos.com created
customresourcedefinition.apiextensions.k8s.io/catalogsources.operators.coreos.com condition met
customresourcedefinition.apiextensions.k8s.io/clusterserviceversions.operators.coreos.com condition met
customresourcedefinition.apiextensions.k8s.io/installplans.operators.coreos.com condition met
customresourcedefinition.apiextensions.k8s.io/olmconfigs.operators.coreos.com condition met
customresourcedefinition.apiextensions.k8s.io/operatorconditions.operators.coreos.com condition met
customresourcedefinition.apiextensions.k8s.io/operatorgroups.operators.coreos.com condition met
customresourcedefinition.apiextensions.k8s.io/operators.operators.coreos.com condition met
customresourcedefinition.apiextensions.k8s.io/subscriptions.operators.coreos.com condition met
namespace/olm created
namespace/operators created
serviceaccount/olm-operator-serviceaccount created
clusterrole.rbac.authorization.k8s.io/system:controller:operator-lifecycle-manager created
clusterrolebinding.rbac.authorization.k8s.io/olm-operator-binding-olm created
olmconfig.operators.coreos.com/cluster created
deployment.apps/olm-operator created
deployment.apps/catalog-operator created
clusterrole.rbac.authorization.k8s.io/aggregate-olm-edit created
clusterrole.rbac.authorization.k8s.io/aggregate-olm-view created
operatorgroup.operators.coreos.com/global-operators created
operatorgroup.operators.coreos.com/olm-operators created
clusterserviceversion.operators.coreos.com/packageserver created
catalogsource.operators.coreos.com/operatorhubio-catalog created
Waiting for deployment "olm-operator" rollout to finish: 0 of 1 updated replicas are available...
deployment "olm-operator" successfully rolled out
deployment "catalog-operator" successfully rolled out
Package server phase: InstallReady
Package server phase: Installing
Package server phase: Succeeded
deployment "packageserver" successfully rolled out


➜ kubectl create -f https://raw.githubusercontent.com/streamnative/charts/master/examples/pulsar-operators/olm-subscription.yaml
subscription.operators.coreos.com/zookeeper-operator created
subscription.operators.coreos.com/bookkeeper-operator created
subscription.operators.coreos.com/pulsar-operator created

➜ kubectl get pods -n operators

NAME                                                      READY   STATUS    RESTARTS        AGE
bookkeeper-operator-controller-manager-68655cf568-mfg7l   2/2     Running   7 (7m26s ago)   30m
pulsar-operator-controller-manager-969694d9-2psjf         2/2     Running   7 (7m17s ago)   30m
zookeeper-operator-controller-manager-55cf6dcdd8-v9xfj    2/2     Running   0               31m
                                                                                                                                                                                                                         
➜ kubectl create ns pulsar
namespace/pulsar created

➜ kubectl create -f https://raw.githubusercontent.com/streamnative/charts/master/examples/pulsar-operators/proxy.yaml
zookeepercluster.zookeeper.streamnative.io/zookeepers created
bookkeepercluster.bookkeeper.streamnative.io/bookies created
pulsarbroker.pulsar.streamnative.io/brokers created
pulsarproxy.pulsar.streamnative.io/proxys created
                                                                                                                                                                                                                        
➜ kubectl get pods -n pulsar   
NAME              READY   STATUS    RESTARTS   AGE
zookeepers-zk-0   1/1     Running   0          27m
zookeepers-zk-1   1/1     Running   0          27m
zookeepers-zk-2   1/1     Running   0          27m

```

I see a lot of restarts of the pulsar and bookkeeper operators, maybe it's the cause?

Sherlock113 commented:
Hi @philippart-s Thank you for the reply.

This is very strange...I tried again and used the same commands as yours and operators, bookies, and brokers are all working properly here. The following are Pods in my OVHcloud Managed K8s cluster:

```
kubectl get pods -A
NAMESPACE     NAME                                                              READY   STATUS      RESTARTS      AGE
kube-system   calico-kube-controllers-74d6b9798d-29xtz                          1/1     Running     0             23m
kube-system   canal-bc8dd                                                       2/2     Running     1 (16m ago)   17m
kube-system   canal-g265z                                                       2/2     Running     1 (17m ago)   18m
kube-system   canal-m9n7b                                                       2/2     Running     1 (17m ago)   18m
kube-system   coredns-78bd9fc67-4tglh                                           1/1     Running     0             17m
kube-system   coredns-78bd9fc67-g4r4z                                           1/1     Running     0             23m
kube-system   kube-dns-autoscaler-7cb9fb9b8b-lphcx                              1/1     Running     0             23m
kube-system   kube-proxy-7m7hz                                                  1/1     Running     0             18m
kube-system   kube-proxy-c7ktl                                                  1/1     Running     0             17m
kube-system   kube-proxy-mg8cv                                                  1/1     Running     0             18m
kube-system   metrics-server-86b4b58bbd-hkpgs                                   1/1     Running     0             22m
kube-system   wormhole-2hvn7                                                    1/1     Running     0             17m
kube-system   wormhole-qdltx                                                    1/1     Running     0             18m
kube-system   wormhole-wplt5                                                    1/1     Running     0             18m
olm           34d69c90f159c510489f2ce3e37e6c030ab7a9262ae071561a3c2ace5028dfx   0/1     Completed   0             11m
olm           3b415493ae775487675a3e6d0f37e68d3d7aefacc55f4265010fc7c9d3lg898   0/1     Completed   0             11m
olm           48396249f58196fd41da66289a2cf9561f40b76cf81146d94652c5322cj6l5p   0/1     Completed   0             11m
olm           catalog-operator-655fb46cd4-cvhl8                                 1/1     Running     0             13m
olm           olm-operator-67fdb4b99d-cc7kp                                     1/1     Running     0             13m
olm           operatorhubio-catalog-hkcf5                                       1/1     Running     0             12m
olm           packageserver-57bf9db95b-8228p                                    1/1     Running     0             12m
olm           packageserver-57bf9db95b-txqzv                                    1/1     Running     0             12m
operators     bookkeeper-operator-controller-manager-76888888c6-jtcgw           2/2     Running     0             10m
operators     pulsar-operator-controller-manager-6c6f7745cc-pxz2n               2/2     Running     0             10m
operators     zookeeper-operator-controller-manager-6747cbb4dd-89bxp            2/2     Running     0             11m
pulsar        bookies-bk-0                                                      1/1     Running     0             5m56s
pulsar        bookies-bk-1                                                      1/1     Running     0             5m56s
pulsar        bookies-bk-2                                                      1/1     Running     0             5m56s
pulsar        bookies-bk-auto-recovery-0                                        1/1     Running     0             5m3s
pulsar        brokers-broker-0                                                  1/1     Running     0             5m50s
pulsar        brokers-broker-1                                                  1/1     Running     0             5m50s
pulsar        proxys-proxy-0                                                    1/1     Running     0             9m3s
pulsar        proxys-proxy-1                                                    1/1     Running     0             9m3s
pulsar        zookeepers-zk-0                                                   1/1     Running     0             9m4s
pulsar        zookeepers-zk-1                                                   1/1     Running     0             9m4s
pulsar        zookeepers-zk-2                                                   1/1     Running     0             9m3s
```

This is my default storage class, which should be the same as yours (I reckon the OVHcloud Managed K8s service has a default one). A default storage class is necessary to ZooKeeper and BookKeeper.

```
kubectl get sc
NAME                              PROVISIONER                RECLAIMPOLICY   VOLUMEBINDINGMODE   ALLOWVOLUMEEXPANSION   AGE
csi-cinder-classic                cinder.csi.openstack.org   Delete          Immediate           true                   24m
csi-cinder-high-speed (default)   cinder.csi.openstack.org   Delete          Immediate           true                   24m
```

One I thing I noticed differently is the node pool. This is my config:

```
kubectl get np
NAME                                            FLAVOR   AUTOSCALED   MONTHLYBILLED   ANTIAFFINITY   DESIRED   CURRENT   UP-TO-DATE   AVAILABLE   MIN   MAX   AGE
nodepool-2cf53f84-xxxx-xxxx-xxxx-f1xxxxxxxxxx   d2-4     false        false           false          3         3         3            3           0     100   26m
```

I did not enable "Anti-affinity" and there are 3 nodes for `CURRENT` and `AVAILABLE` in my config while there was 0 in yours. Does that mean there was no available nodes in your case? Did you use a new OVHcloud Managed K8s cluster or it was shared with other workloads? I am not sure if this could be the reason? 

FYI, here are my nodes info:

```
kubectl get nodes -o wide
NAME                                         STATUS   ROLES    AGE   VERSION   INTERNAL-IP   EXTERNAL-IP     OS-IMAGE             KERNEL-VERSION      CONTAINER-RUNTIME
nodepool-xxxxxxxx-xxxx-xxxx-xx-node-xxxxxx   Ready    <none>   51m   v1.25.9   10.0.0.109    xxx.xx.xx.xx    Ubuntu 22.04.2 LTS   5.15.0-72-generic   containerd://1.6.18
nodepool-xxxxxxxx-xxxx-xxxx-xx-node-xxxxxx   Ready    <none>   51m   v1.25.9   10.0.2.117    xxx.xx.xx.xxx   Ubuntu 22.04.2 LTS   5.15.0-72-generic   containerd://1.6.18
nodepool-xxxxxxxx-xxxx-xxxx-xx-node-xxxxxx   Ready    <none>   52m   v1.25.9   10.0.1.96     xxx.xx.xx.x     Ubuntu 22.04.2 LTS   5.15.0-72-generic   containerd://1.6.18
```

philippart-s commented on pages/platform/kubernetes-k8s/deploying-pulsar-with-streamnative/guide.en-gb.md:
````diff
@@ -0,0 +1,250 @@
+Run the following commands to uninstall OLM:
+
+```bash
+export OLM_RELEASE=<olm-release-version>
````
As the v0.24.0 is used at the beginning of the tutorial I think it's better ti use the same version here ?

```suggestion
export OLM_RELEASE=v0.24.0
```

Sherlock113 replied:
Changed.

philippart-s reviewed (commented):
@Sherlock113 I was able to complete the tutorial 🎉 

My problem was the Kubernetes version, I used a 1.26 and it seems that the operators are not compatible with 1.26.

You mention it in the prerequisites but I miss it, my bad.

I'll check with my PM / PO if it's a problem not using the last recommended version of Kubernetes in an official tutorial.

philippart-s commented:
@Sherlock113 another point : I thing a D2-4 is too small for the proxy version, when I use this kind of cluster `the bookies-bk-auto-recovery-0` was in pending state with not enough memory to run.

Can you update the prerequisites with  6GB of memory as minimum requirement ?

Sherlock113 committed:
Update the doc

Sherlock113 commented:
@philippart-s I updated the doc. Let me know if you have more questions. Thanks.

philippart-s commented:
@Sherlock113 thanks it's perfect!

I'm in contact with with Julien about the Kubernetes version to know if it's possible to have the compatibility with the 1.26 and above.

We'll wait to his answer before validating the PR.

philippart-s commented:
replaced by #4920
